lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -0,0 +1,1499 @@
+                "certificate": self.certificate,
+                "ca": self.ca,
+                "chain": self.chain,
+                "revoked": self.revoked,
+                "expiry_time": self.expiry_time.isoformat(),
+                "expiry_notification_time": self.expiry_notification_time.isoformat()
+                if self.expiry_notification_time
+                else None,
+            }
+        )
+
+
+@dataclass
+class PrivateKey:
+    """This class represents a private key."""
+
+    private_key: str
+    password: str
+
+    def to_json(self) -> str:
+        """Return the object as a JSON string.
+
+        Returns:
+            str: JSON representation of the object
+        """
+        return json.dumps(
+            {
+                "private_key": self.private_key,
+                "password": self.password,
+            }
+        )
+
+
+class CertificateAvailableEvent(EventBase):
+    """Charm Event triggered when a TLS certificate is available."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        certificate: str,
+        certificate_signing_request: str,
+        ca: str,
+        chain: List[str],
+    ):
+        super().__init__(handle)
+        self.certificate = certificate
+        self.certificate_signing_request = certificate_signing_request
+        self.ca = ca
+        self.chain = chain
+
+    def snapshot(self) -> dict:
+        """Return snapshot."""
+        return {
+            "certificate": self.certificate,
+            "certificate_signing_request": self.certificate_signing_request,
+            "ca": self.ca,
+            "chain": self.chain,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.certificate = snapshot["certificate"]
+        self.certificate_signing_request = snapshot["certificate_signing_request"]
+        self.ca = snapshot["ca"]
+        self.chain = snapshot["chain"]
+
+    def chain_as_pem(self) -> str:
+        """Return full certificate chain as a PEM string."""
+        return "\n\n".join(reversed(self.chain))
+
+
+def _get_closest_future_time(
+    expiry_notification_time: datetime, expiry_time: datetime
+) -> datetime:
+    """Return expiry_notification_time if not in the past, otherwise return expiry_time.
+
+    Args:
+        expiry_notification_time (datetime): Notification time of impending expiration
+        expiry_time (datetime): Expiration time
+
+    Returns:
+        datetime: expiry_notification_time if not in the past, expiry_time otherwise
+    """
+    return (
+        expiry_notification_time
+        if datetime.now(timezone.utc) < expiry_notification_time
+        else expiry_time
+    )
+
+
+def csr_has_attributes(  # noqa: C901
+    csr: str,
+    common_name: str,
+    sans_dns: List[str],
+    organization: Optional[str],
+    email_address: Optional[str],
+    country_name: Optional[str],
+    state_or_province_name: Optional[str],
+    locality_name: Optional[str],
+) -> bool:
```

Comment:
I'm a little hesitant about this function, because it seems out of place with the rest of our public interface. It asks the user to pass in the CSR as a string, and I think we (rightly) are using our library objects everywhere else.

I think a better approach here might be to have a CSR string -> CSR object function (`CertificateRequest.from_string(csr)`) and implement `__eq__` in `CertificateRequest`, so it looks more like

```
csr = CertificateRequest.from_string(csr_str)

if (all((
    csr.common_name == ...,
    csr.sans_dns == ...,
    ...
)):
    do_stuff()
```

or

```
if csr == CertificateRequest(
    common_name=...,
    sans_dns=...
):
    do_stuff()
```

Or, if neither of those float your boat, the function could at least exist on the `CertificateRequest` object:

```
csr = CertificateRequest.from_string(csr_str)

if csr.validate(common_name=..., sans_dns=..., ...):
    do_stuff()
```

Reply:
Done